File: updatedashboard.py
#!/usr/bin/env python3
"""
Script to automatically update the API status dashboard by checking
the availability and validity of API credentials
"""
import os
import re
import json
from datetime import datetime
from typing import Dict, Any, Callable, List, Optional, Tuple

# Import local modules
import security
from config import API_KEY, DB_PASSWORD, SECRET_TOKEN

# Constants
DASHBOARD_FILE = os.path.join(os.path.dirname(__file__), 'api_status_dashboard.md')
CREDENTIALS_DIR = 'C:/Documents/credentials/myproject'

# Service definitions with their required credentials
SERVICES = [
    {
        'name': 'Core Authentication API',
        'credential_file': 'API_KEY',
        'validate_fn': security.verify_api_key,
        'credential_var': API_KEY
    },
    {
        'name': 'Database Connection Service',
        'credential_file': 'DB_PASSWORD',
        'validate_fn': security.verify_db_password,
        'credential_var': DB_PASSWORD
    },
    {
        'name': 'Cryptographic Service',
        'credential_file': 'SECRET_TOKEN',
        'validate_fn': lambda token: token and len(token) >= 16,
        'credential_var': SECRET_TOKEN
    },
    {
        'name': 'Mail Service',
        'credential_file': 'MAIL_API_KEY',
        'validate_fn': lambda key: key and len(key) >= 8,
        'credential_var': None
    },
    {
        'name': 'Logging Service',
        'credential_file': 'LOGGING_API_KEY',
        'validate_fn': lambda key: key and len(key) >= 8,
        'credential_var': None
    },
    {
        'name': 'Analytics API',
        'credential_file': 'ANALYTICS_API_KEY',
        'validate_fn': lambda key: key and len(key) >= 8,
        'credential_var': None
    },
    {
        'name': 'Payment Gateway',
        'credential_file': 'PAYMENT_API_KEY',
        'validate_fn': lambda key: key and len(key) >= 8,
        'credential_var': None
    },
    {
        'name': 'External Movie Database API',
        'credential_file': 'MOVIE_DB_API_KEY',
        'validate_fn': lambda key: key and len(key) >= 8,
        'credential_var': None
    },
    {
        'name': 'User Management Service',
        'credential_file': 'USER_MGMT_API_KEY',
        'validate_fn': lambda key: key and len(key) >= 8,
        'credential_var': None
    },
    {
        'name': 'Content Delivery Network',
        'credential_file': 'CDN_API_KEY',
        'validate_fn': lambda key: key and len(key) >= 8,
        'credential_var': None
    }
]


def check_service_status(service: Dict[str, Any]) -> Dict[str, Any]:
    """
    Check if a credential file exists and is valid
    
    Args:
        service: Service to check
        
    Returns:
        Status information dictionary
    """
    # Default status is RED
    status = {
        'color': '🔴 RED',
        'description': 'No codebase embedding created yet',
        'details': 'Pending implementation',
        'has_file': False,
        'is_valid': False
    }
    
    # Check if credential file exists
    file_path = os.path.join(CREDENTIALS_DIR, service['credential_file'])
    try:
        if os.path.exists(file_path):
            status['has_file'] = True
            status['description'] = f"API key found in {CREDENTIALS_DIR}"
            
            # For services integrated in config, check if loaded properly
            if service['credential_var'] is not None:
                is_valid = service['validate_fn'](service['credential_var'])
                if is_valid:
                    status['color'] = '🟢 GREEN'
                    status['details'] = 'Successfully integrated with security module'
                    status['is_valid'] = True
                else:
                    status['color'] = '🟡 ORANGE'
                    status['details'] = 'API key format validation failed'
            else:
                # Check if file content is valid by reading directly
                with open(file_path, 'r') as f:
                    credential = f.read().strip()
                
                try:
                    is_valid = service['validate_fn'](credential)
                    if is_valid:
                        status['color'] = '🟡 ORANGE'
                        status['details'] = 'Credential file exists but not integrated with security module'
                    else:
                        status['color'] = '🟡 ORANGE'
                        status['details'] = 'Credential file exists but has invalid format'
                except Exception as e:
                    status['color'] = '🟡 ORANGE'
                    status['details'] = f'Credential validation error: {str(e)}'
    except Exception as e:
        print(f"Error checking status for {service['name']}: {str(e)}")
    
    return status


def update_dashboard():
    """Update the API status dashboard markdown file"""
    try:
        # Read the current dashboard file
        with open(DASHBOARD_FILE, 'r', encoding='utf-8') as f:
            dashboard_content = f.read()
        
        # Update the timestamp
        now = datetime.now().isoformat()
        updated_content = re.sub(
            r'\*Last Updated: <!-- AUTO_UPDATE_TIMESTAMP -->\*',
            f'*Last Updated: {now}*',
            dashboard_content
        )
        
        # Check service statuses
        service_statuses = [
            (service, check_service_status(service)) 
            for service in SERVICES
        ]
        
        # Update the table in the dashboard
        table_rows = ''
        for service, status in service_statuses:
            table_rows += f"| {service['name']} | {status['color']} | {status['description']} | {status['details']} |\n"
        
        # Replace the table in the dashboard content
        updated_content = re.sub(
            r'\| Service \| Status \| Description \| Integration Details \|\n\|---------|--------|-------------|---------------------\|\n([\s\S]*?)(?=\n## Green Light Services Documentation)',
            f'| Service | Status | Description | Integration Details |\n|---------|--------|-------------|---------------------|\n{table_rows}\n',
            updated_content
        )
        
        # Write the updated dashboard back to file with utf-8 encoding
        with open(DASHBOARD_FILE, 'w', encoding='utf-8') as f:
            f.write(updated_content)
        
        print('API status dashboard updated successfully!')
    except Exception as e:
        print(f'Error updating API status dashboard: {str(e)}')


if __name__ == '__main__':
    update_dashboard() 

File: updatedashboard.js
/**
 * Script to automatically update the API status dashboard by checking
 * the availability and validity of API credentials
 */
const fs = require('fs');
const path = require('path');
const config = require('./config');
const security = require('./security');

// Constants
const DASHBOARD_FILE = path.join(__dirname, 'api_status_dashboard.md');
const CREDENTIALS_DIR = 'C:/Documents/credentials/myproject';

// Service definitions with their required credentials
const SERVICES = [
  {
    name: 'Core Authentication API',
    credentialFile: 'API_KEY',
    validateFn: security.verifyApiKey,
    credentialKey: 'apiKey'
  },
  {
    name: 'Database Connection Service',
    credentialFile: 'DB_PASSWORD',
    validateFn: security.verifyDbPassword,
    credentialKey: 'dbPassword'
  },
  {
    name: 'Cryptographic Service',
    credentialFile: 'SECRET_TOKEN',
    validateFn: (token) => token && token.length >= 16,
    credentialKey: 'secretToken'
  },
  {
    name: 'Mail Service',
    credentialFile: 'MAIL_API_KEY',
    validateFn: (key) => key && key.length >= 8,
    credentialKey: null
  },
  {
    name: 'Logging Service',
    credentialFile: 'LOGGING_API_KEY',
    validateFn: (key) => key && key.length >= 8,
    credentialKey: null
  },
  {
    name: 'Analytics API',
    credentialFile: 'ANALYTICS_API_KEY',
    validateFn: (key) => key && key.length >= 8,
    credentialKey: null
  },
  {
    name: 'Payment Gateway',
    credentialFile: 'PAYMENT_API_KEY',
    validateFn: (key) => key && key.length >= 8,
    credentialKey: null
  },
  {
    name: 'External Movie Database API',
    credentialFile: 'MOVIE_DB_API_KEY',
    validateFn: (key) => key && key.length >= 8,
    credentialKey: null
  },
  {
    name: 'User Management Service',
    credentialFile: 'USER_MGMT_API_KEY',
    validateFn: (key) => key && key.length >= 8,
    credentialKey: null
  },
  {
    name: 'Content Delivery Network',
    credentialFile: 'CDN_API_KEY',
    validateFn: (key) => key && key.length >= 8,
    credentialKey: null
  }
];

/**
 * Check if a credential file exists and is valid
 * @param {Object} service - Service to check
 * @returns {Object} Status information
 */
function checkServiceStatus(service) {
  // Default status is RED
  let status = {
    color: '🔴 RED',
    description: 'No codebase embedding created yet',
    details: 'Pending implementation',
    hasFile: false,
    isValid: false
  };
  
  // Check if credential file exists
  const filePath = path.join(CREDENTIALS_DIR, service.credentialFile);
  try {
    if (fs.existsSync(filePath)) {
      status.hasFile = true;
      status.description = `API key found in ${CREDENTIALS_DIR}`;
      
      // For services integrated in config, check if loaded properly
      if (service.credentialKey && config[service.credentialKey]) {
        const isValid = service.validateFn(config[service.credentialKey]);
        if (isValid) {
          status.color = '🟢 GREEN';
          status.details = 'Successfully integrated with security module';
          status.isValid = true;
        } else {
          status.color = '🟡 ORANGE';
          status.details = 'API key format validation failed';
        }
      } else if (service.credentialKey) {
        // Credential should be in config but isn't
        status.color = '🟡 ORANGE';
        status.details = 'API configured but credentials not properly loaded';
      } else {
        // Check if file content is valid by reading directly
        const credential = fs.readFileSync(filePath, 'utf8').trim();
        const isValid = service.validateFn(credential);
        if (isValid) {
          status.color = '🟡 ORANGE';
          status.details = 'Credential file exists but not integrated with security module';
        } else {
          status.color = '🟡 ORANGE';
          status.details = 'Credential file exists but has invalid format';
        }
      }
    }
  } catch (err) {
    console.error(`Error checking status for ${service.name}:`, err.message);
  }
  
  return status;
}

/**
 * Update the API status dashboard markdown file
 */
function updateDashboard() {
  try {
    // Read the current dashboard file
    const dashboardContent = fs.readFileSync(DASHBOARD_FILE, 'utf8');
    
    // Update the timestamp
    const now = new Date().toISOString();
    let updatedContent = dashboardContent.replace(
      /\*Last Updated: <!-- AUTO_UPDATE_TIMESTAMP -->\*/,
      `*Last Updated: ${now}*`
    );
    
    // Check service statuses
    const serviceStatuses = SERVICES.map(service => {
      const status = checkServiceStatus(service);
      return {
        service,
        status
      };
    });
    
    // Update the table in the dashboard
    let tableRows = '';
    serviceStatuses.forEach(({ service, status }) => {
      tableRows += `| ${service.name} | ${status.color} | ${status.description} | ${status.details} |\n`;
    });
    
    // Replace the table in the dashboard content
    updatedContent = updatedContent.replace(
      /\| Service \| Status \| Description \| Integration Details \|\n\|---------|--------|-------------|---------------------\|\n([\s\S]*?)(?=\n## Green Light Services Documentation)/,
      `| Service | Status | Description | Integration Details |\n|---------|--------|-------------|---------------------|\n${tableRows}\n`
    );
    
    // Write the updated dashboard back to file
    fs.writeFileSync(DASHBOARD_FILE, updatedContent, 'utf8');
    console.log('API status dashboard updated successfully!');
  } catch (err) {
    console.error('Error updating API status dashboard:', err.message);
  }
}

// Run the update when script is executed directly
if (require.main === module) {
  updateDashboard();
}

module.exports = { updateDashboard, checkServiceStatus }; 